src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,585 @@
+     * Usually, the data type of the aggregate can depend on the data type of the corresponding input column.
+     * For example, the sum of ints is an int, while the sum of doubles is an double.
+     * For COUNT(*), the aggregate type is always an integral type, regardless of the input type.
+     * As the input type does not matter and we do not even have an input column,
+     * but the function call expects an input type, we choose Int arbitrarily.
+     */
+    const auto data_type = !column ? DataType::Int : input_table->column_data_type(*column);
+
+    resolve_data_type(
+        data_type, [&, column_index](auto type) { _write_aggregate_output(type, column_index, aggregate.function); });
+
+    ++column_index;
+  }
+
+  auto result_table = std::make_shared<Table>(_output_column_definitions, TableType::Data);
+
+  // Handle empty input table according to the SQL standard
+  // if group by columns exist -> empty result
+  // else -> one row with default values
+  if (input_table->empty()) {
+    if (_groupby_column_ids.empty()) {
+      std::vector<AllTypeVariant> default_values;
+      for (const auto& aggregate : _aggregates) {
+        if (aggregate.function == AggregateFunction::Count || aggregate.function == AggregateFunction::CountDistinct) {
+          default_values.emplace_back(AllTypeVariant{0});
+        } else {
+          default_values.emplace_back(NULL_VALUE);
+        }
+      }
+      result_table->append(default_values);
+    }
+
+    return result_table;
+  }
+
+  /*
+   * NOTE: There are some cases in the TPCH-Queries, where we get tables with more columns than required.
+   * Required columns are all columns that we group by, as well as all columns we aggregate on.
+   * All other columns are superfluous, as they will not be part of the result table (and thus not accessible for later operators).
+   *
+   * This could lead to performance issues as the Sort operator needs to materialize more data.
+   * We believe this to be a minor issue:
+   * It seems that in most cases the optimizer puts a projection before the aggregate operator,
+   * and in cases where it does not, the table size is small.
+   * However, we did not benchmark it, so we cannot prove it.
+   */
+
+  // Sort input table consecutively by the group by columns (stable sort)
+  auto sorted_table = input_table;
+  for (const auto& column_id : _groupby_column_ids) {
+    const auto sorted_wrapper = std::make_shared<TableWrapper>(sorted_table);
+    sorted_wrapper->execute();
+    Sort sort = Sort(sorted_wrapper, column_id);
+    sort.execute();
+    sorted_table = sort.get_output();
+  }
+
+  _output_segments.resize(_aggregates.size() + _groupby_column_ids.size());
+
+  /*
+   * Find all RowIDs where a value in any group by column differs from its predecessor,
+   * as those are exactly the limits of the different groups.
+   * We use a ordered set to "merge" the change locations from multiple columns.
```

Comment:
a**n**. In any case, I find this sentence a bit hard to understand. Maybe: "Everytime a value changes in any of the group by columns, we add the current position to the set.